### PR TITLE
Add fancy indexing support

### LINKED ISSUE #192: Add support for fancy indexing
**Is your feature request related to a problem? Please describe.**

Add the fancy indexing support. It may be helpful for the data transferring between DataFrame and tensor.

**Describe the solution you'd like**

Implement the shuffle map and reduce.

**Additional context**

This requires for the completion of shuffle #134 .

### PR DESCRIPTION
**Pull request by qinxuye**
## What do these changes do?

Fancy indexing is completely supported now, no longer limited to one numpy ndarray.

Three more things need to be done.

- [x] ~~Local cluster seems still have some problem.~~
- [x] ~~Due to the complex logic, the code looks pretty chaotic, refactor is required.~~ Already refactored.
- [x] ~~README can be modified.~~

## Related issue number

Resolves #192

### CODE REVIEW

File: mars/tensor/execution/tests/test_index_execute.py [hunk 1]
```diff
@@ -76,6 +77,107 @@ def testFancyIndexingExecution(self):
...
+        res = self.executor.execute_tensor(arr8, concat=True)
+        np.testing.assert_array_equal(res[0], raw[0, raw_index1, :, raw_index2])
+
+        raw_a = np.random.rand(30, 30)
+        a = tensor(raw_a, chunk_size=(13, 17))
+        b = a.argmax(axis=0)
+        c = a[b, arange(30)]
+        res = self.executor.execute_tensor(c, concat=True)
+
+        self.assertTrue(np.array_equal(res[0], raw_a[raw_a.argmax(axis=0), np.arange(30)]))
```
Comment (sighingnow) [inline mars/tensor/execution/tests/test_index_execute.py]: Why not `np.testing.assert_array_equal`?

Comment (chineking) [reply mars/tensor/execution/tests/test_index_execute.py]: Yes, just to keep the consistency with older codes. I will modify them all later.

File: mars/tensor/core.py [hunk 2]
```diff
@@ -51,6 +51,14 @@ def params(self):
             'index': self.index,
         }
 
+    def __len__(self):
+        try:
+            return self.shape[0]
+        except IndexError:
+            if build_mode().is_build_mode:
+                return 0
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply mars/tensor/core.py]: `IndexError` means it's a scalar. So I think it's fine.

File: mars/tensor/execution/indexing.py [hunk 3]
```diff
@@ -55,6 +63,106 @@ def _index_estimate_size(ctx, chunk):
...
+    nsplits = chunk.op.dest_nsplits
+    axes = chunk.op.axes
+    fancy_index_nsplits = [nsplits[ax] for ax in axes]
+    indexes = ctx[chunk.inputs[0].key]
+    flatten_indexes = indexes.reshape(indexes.shape[0], -1)
+    idx_to_fancy_indexes, idx_to_poses = \
+        split_indexes_into_chunks(fancy_index_nsplits, flatten_indexes, False)
+    for idx in idx_to_fancy_indexes:
+        group_key = ','.join(str(i) for i in idx)
+        ctx[(chunk.key, group_key)] = (idx_to_fancy_indexes[idx], idx_to_poses[idx])
```
[the remark this answers is not in the document]
Comment (qinxuye) [reply mars/tensor/execution/indexing.py]: It's hard to skip chunk for fancy indexing, the empty chunk could be used in the upcoming steps.

**Pull request merged**